viamrobotics/rdk
# [RSDK-7505] fix rtp_passthrough reconfigure

Follow up from https://github.com/viamrobotics/rdk/pull/3723

https://viam.atlassian.net/browse/RSDK-7505

The corresponding [erh/viamrtsp changes](https://github.com/viamrobotics/rdk/pull/3792/files#diff-b3f28beedf03abceca4c114cb88d6dfb3fbdb33ae7c06a3095ca94bfb19abbb2R12-R28)

### How to Review:
1. Start with the rtppassthrough.Source interface changes: https://github.com/viamrobotics/rdk/pull/3792/files#diff-b3f28beedf03abceca4c114cb88d6dfb3fbdb33ae7c06a3095ca94bfb19abbb2R12 and simplifications to stream_subscription.go https://github.com/viamrobotics/rdk/pull/3792/files#diff-944d03d87de4221495114565bb363f37e5a0375a9c6dc0005de11b24638252aa
2. Then read the changes to module.go https://github.com/viamrobotics/rdk/pull/3792/files#diff-9a0a92d26bcd93d88bcfe6d94c11871002ddb11ea827c5f27c75c3238464d079 which use the interface changes to implement more reliable cleanup. 
3. Read the camera client changes (this is the meat of the change): https://github.com/viamrobotics/rdk/pull/3792/files#diff-4e03c7046f4cde2fb3cfabe60f410c7cc0380ec31028edec049f6a37cab65cd8
4.  Review the modmanager tests https://github.com/viamrobotics/rdk/pull/3792/files#diff-32c15f92eba29518fe6de6907202f8fef5cf52320286f9673df9ad5858371a17 which test both the changes to camera/client.go & module.go
5. observe that https://github.com/viamrobotics/rdk/pull/3792/files#diff-db223bcebe63e661c7c3c5b7e0cd20523f08be884e758c78aed2d19ec878f5fa has been reverted back to its state prior to adding experimental RTPPassthrough support. 
6. Review the StreamServer changes https://github.com/viamrobotics/rdk/pull/3792/files#diff-b664e7f5f24e04be1ed075bf7d0f4209eaf9fc88f34eb7acc987ad1693857223 (note no tests needed to change as the behavior is the same if RTPPassthrough is not supported by the camera)
7. Review the new stream/state package https://github.com/viamrobotics/rdk/pull/3792/files#diff-711be8ef6ac79c9940f9d8b2cd13c350370bd347b985236bd4cd5c9253659c10 & associated tests

### Features / Bug Fixes:
1. Support upgrading -> downgrading to / from `rtp_passthrough` and gostream's `GetImage` code path based on what the camera model supports when viam-sever's `stream.AddStream` endpoint is called (prior to this change upgrading required rebooting viam-server).
8. Don't require all peer connections to call `RemoveStream` before broken `rtp_passthrough` video streams recover. A given web browser calling `RemoveStream` -> `AddStream` should be the most required to recover a given web browser's video stream.
9. Fix rtp_passthrough video streams breaking if they are streaming while the module reconfigures
10. Minimize noisy error logging when there is no broken behavior
11. Fix https://viam.atlassian.net/browse/RSDK-7366

### Changes:
1. Change stream/server.go's `AddStream` method to detect if the camera supports `rtp_passthrough`
12. If the camera supports `rtp_passthrough` and this would be the first subscription to that camera, use `rtp_passthrough`
13. If the camera supports `rtp_passthrough` and is already using `rtp_passthrough` continue using `rtp_passthrough` 
14. If the camera supports `rtp_passthrough` but the camera is currently using the gostream `GetImage` code path, stop gostream & start using `rtp_passthrough`
15. If the camera does not support `rtp_passthrough` and this would be the first time we boot gostream and the `GetImage` code path, start that code path
16. If the camera does not support `rtp_passthrough` and the gostream `GetImage` code path is already running, continue using it
17. If at any point the `rtp_passthrough` subscription terminates (i.e. the `Subscription.Context` is cancelled) before RemoveStream is called, re check 2 - 6 & go down the appropriate path.
1. Change signature of `rtppassthrough.Source.SubscribeRTP` to return a `Subscription` rather than a `SubscrioptionID`. A `Subscription` has a `context.Context` which can be check to see if the subscription is no longer active. This is used to detect when the video stream is no longer active due to reasons the subscriber would otherwise not be able to learn about (such as module crashes and model reconfiguration, and errors internal to the camera model such as hardware or network errors). 
4. Removed unused cruft from StreamSubscription (onError callback, etc) 
6. Change camera/client.go's SubcribeRTP function to wait until the 
7. Change module.go to expect `Close` on the camera component which an `rtppassthrough.Source` was derived from to cancel all active subscriptions on that `rtppassthrough.Source`

### Module Compatibility Regression Testing 


#### Tested Compatibility Matrix:

N/A means that the pairing doesn't need to be tested

viam-server v0.23.2 is tested as it doesn't support WebRTC connections

NOTE: Some modules only work on certain platforms. Where the modules didn't work on a given platform I skipped those tests.
|                                                              | ViamServer v0.23.2     | ViamServer with this PR|
| -------------------------------------- | ----------------------|---------------------|
| **Webcam**                                            |             N/A                  |           ✅                           |
| **Fake Camera**                                     |             N/A                  |            ✅                          |
| **erh/viamrtsp 0.0.1**                             |             N/A                  |             ✅                         |
| **erh/viamrtsp with this RDK**              |              ✅                   |           ✅                         |
| **fps camera (python)**                         |              N/A                |                ✅                       |
| **fake camera module with this RDK** |                ✅                  |            ✅                           |

#### Tests (both with & without rtp_passthrough if the model supports rtp_passthrough):
1. Reconfigure the module while video streaming
1. Crash the module while video streaming `kill -3 <pid>`
1. Open / Close video stream in RC card with a single browser a bunch of times
1. Open / Close the video stream in RC card with multiple browsers a bunch of times 
1. Confirm GetImage still works


#### Tested Platforms:
- [x] aarm64 macos
- [x] amd64 linux
- [x] rpi 4 bullseye
- [x] rpi 5 bookworm
